### PR TITLE
 fix(interfaces): Fix thread/exception interface issues

### PR DESCRIPTION
**Pull request by priscilawebdev**
## Description

This Pull Request solves the following problems described in this issue: https://github.com/getsentry/sentry/issues/23870

**The App Only filtering option doesn't display, even if one of the displayed stack traces has "hasSystemFrames" equals true**

_Case 3 - 1, Case 4 - 1, Case 5 - 2_

This was solved in this part of the code 

https://github.com/getsentry/sentry/blob/ab5fc717c3a90ba5e7a47ba32f7742e332c25c9e/src/sentry/static/sentry/app/components/events/interfaces/threads/index.tsx#L42

Now, if the UI displays stack traces of an exception that has at least one stack trace with `hasSystemFrames` equal to` true`, the `App Only` option will also be displayed.

**The UI displays an error message when the event has exceptions without a defined  `threadId`  and a crashed thread without a stack trace**

_Case 6_

The invariants described in this comment https://github.com/getsentry/sentry/pull/23890#issuecomment-780493291 were implemented, solving this issue

### CODE REVIEW

File: src/sentry/static/sentry/app/components/events/interfaces/threads/index.tsx [hunk 1]
```diff
@@ -31,31 +31,31 @@ type Props = {
 } & typeof defaultProps;
 
 type State = {
-  activeThread: Thread;
-  stackView: STACK_VIEW;
   stackType: STACK_TYPE;
   newestFirst: boolean;
+  activeThread?: Thread;
+  stackView?: STACK_VIEW;
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/components/events/interfaces/threads/index.tsx:37]: Updating these types was necessary, as they can be `undefined`

**Comment by jan-auer** [comment]
@priscilawebdev Could you clarify what the final behavior now is of when to show which section? It seems that we're now showing Threads _and_ Exception on one issue details? 

The actual problem is that we're not showing the "In App" toggle in the above case. As soon as the event has threads, the issue details page should _only_ render the Threads section, which renders the exception in place of the crashed thread.

**Comment by untitaker** [comment]
@jan-auer this behavior only triggers if the exception does not have `thread_id`, and at the same time the SDK sends a singular thread containing thread ID. Right now the UI hides the regular exception interface because it sees there will be a threads interface, but then also fails to render the exception interface within the thread as it fails to resolve the thread ID.

We can additionally tell the SDK author to start sending thread ID properly, but I think this is still worth fixing as the UI deals rather poorly with this input by not showing any stacktrace at all.

**Comment by priscilawebdev** [comment]
Thank you for writing a great and detailed explanation @untitaker

**Comment by jan-auer** [comment]
To expand on https://github.com/getsentry/sentry/issues/23870#issuecomment-780452825:

- The exception interface should **never** be rendered if there are threads.
- The UI should render the exception in place of the thread with `crashed: true`. If there is no such thread, the exception should be added to the threads list.
- The SDK should **always** send a `thread_id` in this case. This is actually an SDK bug. However, normalization could compensate for that.

**Comment by priscilawebdev** [comment]
@jan-auer Please what do you mean by this: "If there is no such thread, the exception should be added to the threads list."?

**Comment by jan-auer** [comment]
Updated the description. Does this address your question?

**Comment by priscilawebdev** [comment]
yes Thanks, @jan-auer !

**Comment by priscilawebdev** [comment]
After an offline discussion with Jan, I decided to address the following changes in a follow-up PR:

- In case the `thread_id` is missing, the UI should render the exception on the thread with `crashed: true`.
- If there is neither a `thread_id` nor a thread with `crashed: true`, then we should display the exception either as separate entry in the threads list or on the main thread.

File: src/sentry/static/sentry/app/components/events/interfaces/threads/threadSelector/getThreadStacktrace.tsx [hunk 2]
```diff
@@ -1,24 +1,47 @@
...
+    // In the getThreadException function, when threadIds don't match the activeThread id and the activeThread is 'crashed' and
+    // when at least one exception stack trace is found in the exception values, the function returns the exception data.
+    // When this happen undefined has to be returned here
+    if (thread.crashed) {
+      const exceptionHasAtLeastOneStacktrace = !!threadExceptionValues.find(
+        threadExceptionValue => threadExceptionValue.stacktrace
+      );
+
+      if (!!exceptionHasAtLeastOneStacktrace) {
+        return undefined;
```
Comment (untitaker) [inline src/sentry/static/sentry/app/components/events/interfaces/threads/threadSelector/getThreadStacktrace.tsx:42]: I don't understand this. Isn't this exactly the case where we want to show the exception within the thread? The one we are trying to fix?

Comment (priscilawebdev) [reply src/sentry/static/sentry/app/components/events/interfaces/threads/threadSelector/getThreadStacktrace.tsx:42]: No. The problem we are trying to solve is that when an event has threads without stack traces but exceptions. the UI should display the exceptions.

Comment (untitaker) [reply src/sentry/static/sentry/app/components/events/interfaces/threads/threadSelector/getThreadStacktrace.tsx:42]: oh right, this is getThreadStacktrace 👍

**Review by untitaker — APPROVED** [review]
not super confident but this is probably fine

**Comment by bruno-garcia** [comment]
> If there is neither a `thread_id` nor a thread with `crashed: true`, then we should display the exception either as separate entry in the threads list or on the main thread.

I discussed this with @priscilawebdev on a call looking at the data this looks very broken from the SDK data point of view and starts to get tricky to go in that direction. The Ruby event in question, which isn't clear if it's in production (maybe @st0012 can comment) has a single thread with `crashed=true` and no stack trace, the two threads, one without stack trace.

The current view is actually good for this case since the empty thread isn't really showing up:

image

And the two exceptions (which likely mean are from the same thread) show up one on top of the other.

Trying to fit this into the threads list seems wrong IMO because you can have two exceptions or more exceptions that are different points in time for the same thread and can't really be represented with the thread selector.

Long story short I'm trying to say if this isn't really happening in a production SDK today it might not be worth the effort of messing with the threads/exception components and risk regressions.

**Comment by st0012** [comment]
@bruno-garcia I think `thread_id` should always present in the `exception` when `threads` is added. it was my fault not to add them before (sorry 🙇 ), and I have made a [new release](https://github.com/getsentry/sentry-ruby/blob/master/sentry-ruby/CHANGELOG.md#422) to fix that. so to Ruby users, it's unlikely that this condition will happen:
> If there is neither a thread_id nor a thread with crashed: true

**Comment by bruno-garcia** [comment]
> @bruno-garcia I think `thread_id` should always present in the `exception` when `threads` is added. it was my fault not to add them before (sorry 🙇 ), and I have made a [new release](https://github.com/getsentry/sentry-ruby/blob/master/sentry-ruby/CHANGELOG.md?rgh-link-date=2021-02-18T15%3A25%3A00Z#422) to fix that. so to Ruby users, it's unlikely that this condition will happen:
> 
> > If there is neither a thread_id nor a thread with crashed: true
> 
> (btw I think the updated view looks pretty good 👍)

Right, I just wonder if this was the behavior on the SDK since the Raven times or it was just a minor version out for a few days. In the latter case it we have less need to try to compensate for this in the UI and it's just a matter of waiting for that version to fade away/steer users to upgrade the SDK.

**Comment by st0012** [comment]
@bruno-garcia it was out for about 2 weeks with a minor version bump (affected versions: `4.2.0` and `4.2.1`). so the adopted users should be relatively few, you may be able to get an exact number by querying from the database with the versions above. also, in the exception events, the threads always had `crashed: true` set, just without the `thread_id`. 

so I don't think a further step is needed.

**Comment by priscilawebdev** [comment]
@jan-auer I implemented your suggestions, but in my opinion, we shouldn't make these changes now, as everything is very fragile.

image

**Comment by jan-auer** [comment]
Please check for the updated conditions in https://github.com/getsentry/sentry/pull/23890#issuecomment-780493291.

File: src/sentry/static/sentry/app/components/events/interfaces/threads/threadSelector/getThreadStacktrace.tsx [hunk 3]
```diff
@@ -1,24 +1,48 @@
...
-          rv = singleExc.rawStacktrace;
-        }
-      }
+function getThreadStacktrace(
+  event: Event,
+  raw: boolean,
+  thread?: Thread,
+  exception?: Required<ExceptionType>,
+  checkException?: boolean
+) {
```
Comment (scttcper) [inline src/sentry/static/sentry/app/components/events/interfaces/threads/threadSelector/getThreadStacktrace.tsx:13]: is it possible to add an explicit return type, seems like there's a few options

Comment (priscilawebdev) [reply src/sentry/static/sentry/app/components/events/interfaces/threads/threadSelector/getThreadStacktrace.tsx:13]: the only options are `undefined` and `StacktraceType`, but I can add an explicit return type 😉

**Review by markstory — APPROVED** [review]
JS code looks good to me. I didn't look at the new fixtures, but the snapshots look good.

**Review by jan-auer — APPROVED** [review]
Fixtures look good, and verified the behavior in review with @priscilawebdev
